dhall/src/Dhall/Deriving.hs

```diff
@@ -0,0 +1,515 @@
+{-# LANGUAGE AllowAmbiguousTypes  #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE PolyKinds            #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-| Newtypes for writing customizable 'FromDhall' and 'ToDhall' instances
+    through the DerivingVia strategy.
+
+    Inspired by Matt Parson's blog post
+    [Mirror Mirror: Reflection and Encoding Via](https://www.parsonsmatt.org/2020/02/04/mirror_mirror.html),
+    but applied to Dhall instead of JSON.
+-}
+
+module Dhall.Deriving
+  (
+    -- * Introduction
+    -- $introduction
+
+    -- * Writing FromDhall instances by hand
+    -- $instancesByHand
```

Comment:
These motivating sections are very good, but also quite long. I fear that some users with short attention spans (like myself) might give up before they find out how _this module_ can be used.

So maybe put `$derivingVia` at the top, move the other sections to the bottom of the page, but reference them from that top section? Just an idea…

Reply:
yeah I wasn't sure how to structure it. I'll look into that later today, but your suggestion sounds good

Reply:
- [ ] reorder documentation sections

Reply:
I tried your suggestion but I didn't like the result, since (surprisingly) the Haddock story for linking to headings or documentation chunks isn't there. Instead, I linked to the `$derivingVia` section from the header. Let me know what you think.

Reply:
Fair enough! Should be easy to change if we later come up with a better idea!